compiler/rustc_resolve/src/late.rs

```diff
@@ -3316,14 +3318,36 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
             self.visit_ty(&qself.ty);
         }
         self.visit_path(&delegation.path, delegation.id);
+        let last_segment = delegation.path.segments.last().unwrap();
+
+        // Saving traits for a `MethodCall` that has not yet been generated.
+        // Traits found in the path are also considered visible:
+        //
+        // impl Trait for Type {
+        //    reuse inner::TraitFoo::*; // OK, even `TraitFoo` is not in scope.
+        // }
+        let mut traits = self.traits_in_scope(last_segment.ident, ValueNS);
+        for segment in &delegation.path.segments {
+            if let Some(partial_res) = self.r.partial_res_map.get(&segment.id) {
+                let res_id = partial_res.expect_full_res().def_id();
```

Comment:
Could you check if something like `reuse u8;` panics here because it doesn't have a `DefId`?
Unexpected resolutions should be turned into `Res::Err`s, but `Res::Err` can also panic here.